modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/transaction/TransactionBindDelegate.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.transaction;
+
+import static java.lang.String.format;
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.extension.api.util.NameUtils.getComponentModelTypeName;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.connection.ConnectionHandler;
+import org.mule.runtime.api.meta.model.ComponentModel;
+import org.mule.runtime.api.meta.model.ExtensionModel;
+import org.mule.runtime.api.tx.TransactionException;
+import org.mule.runtime.core.api.transaction.Transaction;
+import org.mule.runtime.core.api.transaction.TransactionConfig;
+import org.mule.runtime.extension.api.connectivity.TransactionalConnection;
+import org.mule.runtime.extension.api.connectivity.XATransactionalConnection;
+
+import java.util.function.Supplier;
+
+/**
+ * Binds a connection to a given transaction and returns the {@link ConnectionHandler} that has been bound to it.
+ *
+ * @since 4.0
+ */
+public class TransactionBindDelegate {
+
+  /**
+   * @param componentModel model of the component to enter to the transaction
+   * @param extensionModel model of the current extension
+   * @param transactionConfig given transaction config
+   * @param txKey the transaction key
+   * @param currentTx current transaction
+   * @param connectionHandlerSupplier {@link Supplier} to get the {@link ConnectionHandler} of the current component
+   * @return The {@link ConnectionHandler} that has be bound to the transaction.
+   * @throws ConnectionException if a problem occurred retrieving the {@link ConnectionHandler}
+   * @throws TransactionException if the connection could not be bound to the current transaction
+   */
+  public <T extends TransactionalConnection> ConnectionHandler<T> bindResource(ComponentModel componentModel,
+                                                                               ExtensionModel extensionModel,
+                                                                               TransactionConfig transactionConfig,
+                                                                               ExtensionTransactionKey txKey,
+                                                                               Transaction currentTx,
+                                                                               ConnectionSupplier<ConnectionHandler<T>> connectionHandlerSupplier)
+      throws ConnectionException, TransactionException {
+    if (currentTx != null) {
```

Comment:
I don't think we should call a "bind resource" over a null Transaction. Seems like this should be handled in the outside